### PR TITLE
Feature/expose getpublickey

### LINKED ISSUE #4: Expose getPublicKey rpc api
- expose getPublicKey rpc api (without params)
- method should returned public key of generated account
- display that public key after connect in index.html

### PR DESCRIPTION
**Pull request by MakMuftic**
Closes #4 

@mpetrunic I had a problem with stubbing exported functions (example `generateKeyPairFromSeed` in `keys.ts`), so I know my tests are not ideal. I am still looking into this, but I think this is sufficient for now.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -25,43 +25,61 @@ <h1>Hello Plugins!</h1>
...
 
-    async function connect () {
-      await ethereum.send({
-        method: 'wallet_requestPermissions',
-        params: [{
-          [snapId]: {}
-        }]
-      })
-    }
+    fetchPublicKey().then((fetched) => {
```
Comment (mpetrunic) [inline index.html]: Why do you need this? You are calling this after connect

File: src/crypto/error.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+export class InvalidSeedError extends Error {
+  public static INVALID_SEED_MESSAGE = "Provided seed string needs to be longer or equal than 32 characters";
+  constructor() {
+    super();
+    this.name = "InvalidSeedError";
+    this.message = InvalidSeedError.INVALID_SEED_MESSAGE;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any,@typescript-eslint/consistent-type-assertions
+    this.stack = (<any> new Error()).stack;
+  }
+}
```
Comment (mpetrunic) [inline src/crypto/error.ts]: https://github.com/MetaMask/eth-json-rpc-errors

File: src/rpc/getPublicKey.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import {Wallet} from "../interfaces";
+import {generateKeys} from "../crypto/keys";
+import {iterableToHexString} from "../util";
+
+export async function getPublicKey(wallet: Wallet): Promise<string> {
+  const keyPairState = wallet.getPluginState();
+  if (keyPairState != null && keyPairState.polkadot.account != null) {
```
Comment (mpetrunic) [inline src/rpc/getPublicKey.ts]: `!keyPairState && !keyPairState.polkadot.account` so you are chcking both null and undefined

File: src/util.ts [hunk 4]
```diff
@@ -1,17 +1,26 @@
...
-  const seed = convertStringTo8UintArray(seedString.substr(0, 32));
-  return nacl.sign.keyPair.fromSeed(seed);
+// define byte to hex mapping
+const byteToHex: string[] = [];
+for (let n = 0; n <= 0xff; ++n) {
+  const hexOctet = String("00" + n.toString(16)).slice(-2);
+  byteToHex.push(hexOctet);
+}
+
+export function iterableToHexString(arrayBuffer: Iterable<number>): string {
```
Comment (mpetrunic) [inline src/util.ts]: This method works for Uint8Array only, not all number arrays. Just call it `toHexString`

File: src/util.ts [hunk 5]
```diff
@@ -1,17 +1,26 @@
...
+  const buff = new Uint8Array(arrayBuffer);
+  const hexOctets = new Array(buff.length);
+
+  for (let i = 0; i < buff.length; ++i)
+    hexOctets[i] = byteToHex[buff[i]];
+
+  return hexOctets.join("");
```
Comment (mpetrunic) [inline src/util.ts]: 0x prefix

File: src/util.ts [hunk 6]
```diff
@@ -1,17 +1,26 @@
-import nacl from "tweetnacl";
-import {KeyPairState} from "./interfaces";
 
-function convertStringTo8UintArray(string: string): Uint8Array {
+export function convertStringTo8UintArray(string: string): Uint8Array {
```
Comment (mpetrunic) [inline src/util.ts]: handle possible 0x prefix and check if string.length%2 === 0

Comment (mpetrunic) [reply src/util.ts]: rename to fromHexString

File: src/util.ts [hunk 7]
```diff
@@ -1,17 +1,26 @@
-import nacl from "tweetnacl";
```
Comment (mpetrunic) [inline src/util.ts]: rename this file to utils/hex.ts

File: test/unit/crypto/keys.test.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import sinon from "sinon";
+import {generateKeys} from "../../../src/crypto/keys";
+import {InvalidSeedError} from "../../../src/crypto/error";
+
+chai.use(sinonChai);
+
+describe('generateKeys', () => {
+
+  const sandbox = sinon.createSandbox();
+  const walletStub = {} as Wallet;
```
Comment (mpetrunic) [inline test/unit/crypto/keys.test.ts]: in separate file implement MockWallet implement Wallet and implement every method like this:
`public updatePluginState = sinon.stub();`

File: test/unit/crypto/keys.test.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+  it('should generate valid keypair from app key', async () => {
+    // stubs
+    walletStub.getAppKey = sandbox.stub().returns("abasddsa12ssavdasfdas2easdfa21sa");
+    walletStub.updatePluginState = sandbox.stub();
+    // tested method
+    const result = await generateKeys(walletStub);
+    // assertions
+    expect(walletStub.getAppKey).to.have.been.calledOnce;
+    expect(walletStub.updatePluginState).to.have.been.calledOnce;
```
Comment (mpetrunic) [inline test/unit/crypto/keys.test.ts]: check params of method call

File: index.html (hunk 1 above)
Comment (MakMuftic) [reply index.html]: In case you already connected snap and have saved public key in a state.

File: src/crypto/error.ts (hunk 2 above)
Comment (MakMuftic) [reply src/crypto/error.ts]: @mpetrunic I will be on the lookout for _typescript migration PR_ to finish and then migrate our errors to use this.

Comment (mpetrunic) [reply src/crypto/error.ts]: it already has typescript definitions

Comment (mpetrunic) [reply src/crypto/error.ts]: but now that im thinking of it, it might make more sense to throw substrate rpc errors

File: src/util/hex.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
+
+export function fromHexString(string: string): Uint8Array {
```
Comment (MakMuftic) [inline src/util/hex.ts]: @mpetrunic just check hex functions.

File: src/util/hex.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  let s = string;
+  // clear prefix
+  if (string.startsWith("0x")) {
+    s = string.slice(2);
+  }
+  // check string length
+  if (s.length % 2 !== 0) {
+    throw new Error("Invalid hex string, length not even number.");
+  }
+  if (!s.match(/^[0-9a-f]+$/)) {
```
Comment (mpetrunic) [inline src/util/hex.ts]: uppercase letters are fine as well

Comment (MakMuftic) [reply src/util/hex.ts]: @mpetrunic my bad :disappointed:  I added an additional test.

**Pull request merged**